In [5]:
import math
import pandas as pd

In [9]:
df = pd.read_csv('yandex_data.csv')
df.head()

x   F                   
0  9.481538             394.853787
1  6.878546             238.295702
2  9.436524             393.053669
3  3.572694              55.666155
4  1.079596              13.537515

In [27]:
def f(a, b, c, x):
    return (a * math.sin(x) + b * math.log(x)) ** 2 + c * (x **2)

In [28]:
def error_da(f, a, b, c, x, F):
    da = 0.0001
    f1 = f(a, b, c, x)
    f2 = f(a + da, b, c, x)
    e1 = (F - f1) ** 2
    e2 = (F - f2) ** 2
    return (e2 - e1) / da

def error_db(f, a, b, c, x, F):
    db = 0.0001
    f1 = f(a, b, c, x)
    f2 = f(a, b + db, c, x)
    e1 = (F - f1) ** 2
    e2 = (F - f2) ** 2
    return (e2 - e1) / db

def error_dc(f, a, b, c, x, F):
    dc = 0.0001
    f1 = f(a, b, c, x)
    f2 = f(a, b, c + dc, x)
    e1 = (F - f1) ** 2
    e2 = (F - f2) ** 2
    return (e2 - e1) / dc

In [36]:
a, b, c = 1, 1, 1
learning_rate = 0.0001
for j in range(100):
    for i, row in df.iterrows():
        x, F = row[0], row[1]
        a -= learning_rate * error_da(f, a, b, c, x, F)
        b -= learning_rate * error_db(f, a, b, c, x, F)
        c -= learning_rate * error_dc(f, a, b, c, x, F)
    
print(a, b, c)

3.140978897774492 2.719421180940797 4.000377257048


In [31]:
def error_da2(f, a, b, c, x, F):
    da = 0.0001
    f1 = f(a, b, c, x)
    f2 = f(a + da, b, c, x)
    e1 = abs(F - f1)
    e2 = abs(F - f2)
    return (e2 - e1) / da

def error_db2(f, a, b, c, x, F):
    db = 0.0001
    f1 = f(a, b, c, x)
    f2 = f(a, b + db, c, x)
    e1 = abs(F - f1)
    e2 = abs(F - f2)
    return (e2 - e1) / db

def error_dc2(f, a, b, c, x, F):
    dc = 0.0001
    f1 = f(a, b, c, x)
    f2 = f(a, b, c + dc, x)
    e1 = abs(F - f1)
    e2 = abs(F - f2)
    return (e2 - e1) / dc

In [40]:
a, b, c = 1, 1, 1
learning_rate = 0.001
for j in range(1):
    for i, row in df.iterrows():
        x, F = row[0], row[1]
        a -= learning_rate * error_da2(f, a, b, c, x, F)
        b -= learning_rate * error_db2(f, a, b, c, x, F)
        c -= learning_rate * error_dc2(f, a, b, c, x, F)
    
print(a, b, c)

3.1690300157412934 2.731369060315344 4.019544481479311


a is a PI number 3.14

b is an Eiler number 2.71

and c is ~ 4